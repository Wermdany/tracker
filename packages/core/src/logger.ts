export interface LoggerConfig {
  isLogger: boolean;
}

export interface Logger {
  warn: (...args: any[]) => void;
  info: (...args: any[]) => void;
  error: (...args: any[]) => void;
  throwError: (base: Error | string) => never;
}

export function createLogger<N extends string>(name: N, options?: Partial<LoggerConfig>): Logger {
  const { isLogger = true } = options || {};

  const genPrefix = () => {
    return name ? "[".concat(__NAME__, "/", name, "]") : "[".concat(__NAME__, "]");
  };

  return {
    warn(...args: any[]) {
      isLogger && console.warn(genPrefix(), ...args);
    },
    info(...args: any[]) {
      isLogger && console.info(genPrefix(), ...args);
    },
    error(...args: any[]) {
      isLogger && console.error(genPrefix(), ...args);
    },
    throwError(base: Error | string): never {
      if (!(base instanceof Error)) {
        base = new Error(base);
        Object.defineProperty(base, "name", {
          value: "".concat(titleToUpperCase(__NAME__), titleToUpperCase(name), "Error")
        });
      }
      throw base;
    }
  };
}

function titleToUpperCase(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
