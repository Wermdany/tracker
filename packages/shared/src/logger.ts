export interface Logger {
  bold(msg: string): string;
  info(...args: any[]): void;
  log(...args: any[]): void;
  warn(...args: any[]): void;
  error(...args: any[]): void;
  debug(...args: any[]): void;
  msg(...args: string[]): string;
}

type LogType = "info" | "warn" | "error";

export interface LoggerOptions {
  prefix: string;
}

interface MethodsOptions {
  prefix?: string;
  type?: LogType;
}

export function createLogger(module?: string, options: Partial<LoggerOptions> = {}): Logger {
  const { prefix = "@tracker" } = options;

  const bold = (word: string) => {
    return `\x1b[1m${word}\x1b[22m`;
  };

  const pre = module ? bold(`[${prefix}/${module}] `) : bold(`[${prefix}] `);

  const methods = (options: MethodsOptions, ...args: string[]) => {
    const { type, prefix = pre } = options;
    if (!type) {
      return [prefix].concat(args).join(" ");
    }
    console[type](prefix, ...args);
  };

  const logger: Logger = {
    bold,
    log(...args) {
      methods({ type: "info" }, ...args);
    },
    info(...args) {
      methods({ type: "info" }, ...args);
    },
    warn(...args) {
      methods({ type: "warn" }, ...args);
    },
    error(...args) {
      methods({ type: "error" }, ...args);
    },
    debug(...args) {
      const debug = "\x1B[45m\x1B[97m DEBUG \x1B[39m\x1B[49m";
      methods({ type: "info", prefix: debug }, ...args);
    },
    msg(...args) {
      return methods({}, ...args)!;
    }
  };

  return logger;
}
