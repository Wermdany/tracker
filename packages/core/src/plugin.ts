import type { Logger } from "./logger";
import type { TrackerCore } from "./core";
import type { TrackerCoreConfig } from "./config";

import { createLogger } from "./logger";

export interface NormalPlugin {
  name: string;
}

export interface FunctionPlugins {
  // 同步 顺序执行
  setup: (this: PluginContext, initConfig: TrackerCoreConfig) => void;
  // 同步 顺序执行
  init: (this: PluginContext, ctx: TrackerCore) => void;
  // 同步 顺序执行
  report: (this: PluginContext, event: string, record: Record<string, any>) => void;
  // 同步 循环合并
  build: (this: PluginContext, event: string, record: Record<string, any>) => Record<string, any>;
  // 同步 熔断执行
  beforeSend: (this: PluginContext, event: string, build: Record<string, any>) => boolean | void | null;
  // 同步 熔断执行
  send: (this: PluginContext, event: string, build: Record<string, any>) => boolean | void | null;
  // 异步 并行执行
  beforeDestroy: (this: PluginContext, ctx: TrackerCore) => any;
  // 同步 顺序执行
  destroy: (this: PluginContext, ctx: TrackerCore) => void;
}

type ObjectHook<T, O = {}> = T | ({ handler: T; order?: "pre" | "post" | null } & O);

type MakeAsync<Fn> = Fn extends (this: infer This, ...args: infer Args) => infer Return
  ? (this: This, ...args: Args) => Return | Promise<Return>
  : never;

export type AsyncPluginHooks = "beforeDestroy";

export type SyncPluginHooks = Exclude<keyof FunctionPlugins, AsyncPluginHooks>;
// 按照顺序执行
export type SequentialPluginHooks = "setup" | "init" | "report" | "destroy";
// 同时执行
export type ParallelPluginHooks = "beforeDestroy";
// 只要有一个错误，则直接返回
export type BailPluginHooks = "beforeSend" | "send";
// 需要按照顺序循环合并值
export type ChainMergeHooks = Exclude<
  keyof FunctionPlugins,
  SequentialPluginHooks | ParallelPluginHooks | BailPluginHooks
>;

export type PluginHooks = {
  [K in keyof FunctionPlugins]: ObjectHook<
    K extends AsyncPluginHooks ? MakeAsync<FunctionPlugins[K]> : FunctionPlugins[K]
  >;
};

export type TrackerPlugin = Partial<PluginHooks> & NormalPlugin;

export function PluginDriver(plugins: TrackerPlugin[], options: Partial<TrackerCoreConfig>, logger: Logger) {
  const sortedPlugins: Map<keyof FunctionPlugins, TrackerPlugin[]> = new Map();
  const pluginContexts: ReadonlyMap<TrackerPlugin, PluginContext> = new Map(
    plugins.map(plugin => [plugin, createContext(plugin, options)])
  );

  async function runHook<H extends AsyncPluginHooks>(
    hookName: H,
    args: Parameters<FunctionPlugins[H]>,
    plugin: TrackerPlugin
  ): Promise<ReturnType<FunctionPlugins[H]>>;

  async function runHook<H extends AsyncPluginHooks>(hookName: H, args: unknown[], plugin: TrackerPlugin) {
    const hook = plugin[hookName]!;

    const handler = typeof hook === "object" ? hook.handler : hook;

    const context = pluginContexts.get(plugin)!;

    return Promise.resolve()
      .then(() => {
        const hookResult = (handler as Function).apply(context, args);

        if (!hookResult?.then) {
          return hookResult;
        }
        return hookResult;
      })
      .catch(err => logger.throwError(err));
  }

  function runHookSync<H extends SyncPluginHooks>(
    hookName: H,
    args: Parameters<FunctionPlugins[H]>,
    plugin: TrackerPlugin
  ): ReturnType<FunctionPlugins[H]> {
    const hook = plugin[hookName]!;
    const handler = typeof hook === "object" ? hook.handler : hook;

    const context = pluginContexts.get(plugin)!;

    try {
      return (handler as Function).apply(context, args);
    } catch (err: any) {
      return logger.throwError(err);
    }
  }

  function getSortedValidatePlugins(
    hookName: keyof FunctionPlugins,
    plugins: TrackerPlugin[],
    validateHandler = validateFunctionPluginHandler
  ) {
    const pre: TrackerPlugin[] = [];
    const normal: TrackerPlugin[] = [];
    const post: TrackerPlugin[] = [];

    for (const plugin of plugins) {
      const hook = plugin[hookName];
      if (hook) {
        if (typeof hook === "object") {
          validateHandler(hook.handler, hookName, plugin);
          if (hook.order == "pre") {
            pre.push(plugin);
            continue;
          }
          if (hook.order == "post") {
            post.push(plugin);
            continue;
          }
          normal.push(plugin);
        } else {
          validateHandler(hook, hookName, plugin);
          normal.push(plugin);
        }
      }
    }

    return pre.concat(normal, post);
  }

  function validateFunctionPluginHandler(handler: unknown, hookName: string, plugin: TrackerPlugin) {
    if (typeof handler !== "function") {
      logger.throwError(
        `Error running plugin hook "${hookName}" for plugin "${plugin.name}", expected a function hook or an object with a "handler" function.`
      );
    }
  }

  function getSortedPlugins(hookName: keyof FunctionPlugins) {
    return getOrCreate(sortedPlugins, hookName, () => getSortedValidatePlugins(hookName, plugins));
  }

  return {
    runHook,
    runHookSync,
    getSortedValidatePlugins,
    async hookParallel<H extends ParallelPluginHooks & AsyncPluginHooks>(
      hookName: H,
      args: Parameters<FunctionPlugins[H]>
    ): Promise<void> {
      const parallelPromises: Promise<unknown>[] = [];

      for (const plugin of getSortedPlugins(hookName)) {
        parallelPromises.push(runHook(hookName, args, plugin));
      }

      await Promise.all(parallelPromises);
    },
    hookSequentialSync<H extends SequentialPluginHooks & SyncPluginHooks>(
      hookName: H,
      args: Parameters<FunctionPlugins[H]>
    ) {
      for (const plugin of getSortedPlugins(hookName)) {
        runHookSync(hookName, args, plugin);
      }
    },
    hookBail<H extends BailPluginHooks & SyncPluginHooks>(
      hookName: H,
      args: Parameters<FunctionPlugins[H]>
    ): boolean {
      for (const plugin of getSortedPlugins(hookName)) {
        const result = runHookSync(hookName, args, plugin);
        // 只有在明确返回 false 才会熔断，这个是考虑避免用户写过于复杂判断，没有 return 而导致熔断
        if (result === false) {
          return false;
        }
      }
      return true;
    },
    hookChainMergeSync<H extends ChainMergeHooks & SyncPluginHooks>(
      hookName: H,
      args: Parameters<FunctionPlugins[H]>,
      initialValue: Record<string, any>
    ) {
      let returnValue = initialValue;
      for (const plugin of getSortedPlugins(hookName)) {
        const result = runHookSync(hookName, args, plugin);
        returnValue = mergeConfig(returnValue, result);
      }
      return returnValue;
    }
  };
}

interface ContextMeta {
  version: string;
}

export interface PluginContext {
  logger: Logger;
  meta: ContextMeta;
}

export function createContext(plugin: TrackerPlugin, options: Partial<TrackerCoreConfig>): PluginContext {
  return {
    logger: createLogger(plugin.name, options),
    meta: {
      version: __VERSION__
    }
  };
}

function getOrCreate<K, V>(plugins: Map<K, V>, key: K, init: () => V): V {
  const existing = plugins.get(key);
  if (existing) {
    return existing;
  }
  const value = init();
  plugins.set(key, value);
  return value;
}

export function mergeConfig(defaults: Record<string, any>, overrides: Record<string, any>) {
  const merged: Record<string, any> = { ...defaults };

  for (const key in overrides) {
    if (!Object.prototype.hasOwnProperty.call(overrides, key)) {
      continue;
    }

    const value = overrides[key];

    if (value == null) {
      continue;
    }

    const existing = merged[key];

    if (existing == null) {
      merged[key] = value;
      continue;
    }

    if (Array.isArray(existing) || Array.isArray(value)) {
      merged[key] = [...toArray(existing ?? []), ...toArray(value ?? [])];
      continue;
    }

    if (isObject(existing) && isObject(value)) {
      merged[key] = mergeConfig(existing, value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

function toArray<T>(target: T | T[]) {
  return Array.isArray(target) ? target : [target];
}

function isObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === "[object Object]";
}
