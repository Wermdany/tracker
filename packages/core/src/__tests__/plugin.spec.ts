import type { FunctionPlugins, TrackerPlugin } from "../plugin";
import type { TrackerCore } from "../core";
import type { TrackerCoreConfig } from "../config";

import { mergeConfig, PluginDriver } from "../plugin";
import { createLogger } from "../logger";

const logger = createLogger("core");

const ctx = {} as TrackerCore;

const coreConfig = {} as TrackerCoreConfig;

describe("test PluginDriver utils", () => {
  it("test mergeConfig", () => {
    const defaults = { a: { b: { c: 2 }, d: [1, 2] }, e: 1 };
    // 必须重建一个新的不能直接添加属性
    expect(mergeConfig(defaults, {})).not.toBe(defaults);
    // 忽略合并上的虚值
    expect(mergeConfig(defaults, { d: null })).toEqual(defaults);
    //忽略原本的虚值
    expect(mergeConfig({ f: undefined }, defaults)).toEqual(defaults);
    // 复杂的合并策略
    expect(mergeConfig(defaults, { a: { b: { c: 3 }, d: 2 }, e: null })).toEqual({
      a: { b: { c: 3 }, d: [1, 2, 2] },
      e: 1
    });
  });
});

describe("test PluginDriver inside Api", () => {
  const { getSortedValidatePlugins, runHookSync, runHook } = PluginDriver([], {}, logger);

  it("runHook success", () => {
    expect(
      runHook("beforeDestroy", [ctx], {
        name: "test",
        beforeDestroy: async () => {
          await Promise.resolve();
          return true;
        }
      })
    ).resolves.toBe(true);
  });

  it("runHook error", () => {
    expect(
      runHook("beforeDestroy", [ctx], {
        name: "test",
        beforeDestroy: () => {
          return Promise.reject("error");
        }
      })
    ).rejects.toThrowError("error");
  });

  it("runHookSync success", () => {
    const record = runHookSync("build", ["test", { a: 1 }], {
      name: "test",
      build: (event, record) => ({ event, record })
    });

    expect(record).toEqual({ event: "test", record: { a: 1 } });
  });

  it("runHookSync error", () => {
    expect(() =>
      runHookSync("build", ["test", { a: 1 }], {
        name: "test",
        build: {
          handler: () => {
            // @ts-ignore
            record;
            return { a: 1 };
          }
        }
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it("getSortedValidatePlugins", () => {
    const sorted = getSortedValidatePlugins("init", [
      {
        name: "1",
        init: {
          order: "post",
          handler: () => ({})
        }
      },
      {
        name: "2",
        init: () => ({})
      },
      {
        name: "3",
        init: {
          order: "pre",
          handler: () => ({})
        }
      },
      {
        name: "4",
        init: {
          handler: () => ({})
        }
      }
    ]);

    expect(sorted.map(item => item.name).join("->")).toBe("3->2->4->1");

    expect(() =>
      getSortedValidatePlugins("build", [
        {
          name: "test",
          // @ts-ignore
          build: 1
        }
      ])
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("test PluginDriver core", () => {
  it("init & prototype", () => {
    const pd = PluginDriver(
      [
        {
          name: "testPlugin",
          setup() {
            this.logger.info(this.meta.version);
          }
        }
      ],
      {},
      logger
    );

    expect(pd.hookSequentialSync("setup", [coreConfig])).toHaveConsoleInfo("testPlugin", __VERSION__);
  });

  it("hookParallel Api", () => {
    const flat = [false, false];

    const plugins = genPlugins(
      "beforeDestroy",
      flat.map((_, index) => async () => {
        await Promise.resolve();
        flat[index] = true;
      })
    );
    const pd = PluginDriver(plugins, {}, logger);

    pd.hookParallel("beforeDestroy", [ctx]).then(() => {
      expect(flat).toEqual([true, true]);
    });
  });

  it("hookSequentialSync Api", () => {
    const flat: number[] = [];
    const resolve = [1, 2, 3, 4];
    const plugins = genPlugins(
      "setup",
      resolve.map(item => () => {
        flat.push(item);
      })
    );
    const pd = PluginDriver(plugins, {}, logger);
    pd.hookSequentialSync("setup", [coreConfig]);

    expect(flat).toEqual(resolve);
  });

  it("hookBail Api", () => {
    let flat = -1;
    const resolve = [1, 2, 3, 4];
    const plugins = genPlugins(
      "beforeSend",
      resolve.map(item => () => {
        flat = item;
        return item !== 3;
      })
    );
    const pd = PluginDriver(plugins, {}, logger);
    pd.hookBail("beforeSend", ["test", {}]);

    expect(flat).toEqual(3);
  });

  it("hookChainMergeSync Api", () => {
    const resolve = [1, 2, 3, 4];
    const plugins = genPlugins(
      "build",
      resolve.map(item => () => {
        return {
          value: item
        };
      })
    );
    const pd = PluginDriver(plugins, {}, logger);
    expect(pd.hookChainMergeSync("build", ["test", {}], {})).toEqual({ value: 4 });
  });
});

function genPlugins(hook: keyof FunctionPlugins, handlers: Array<(...args: any[]) => any>): TrackerPlugin[] {
  return handlers.map((handler, index) => ({
    name: "test" + index,
    [hook]: handler
  }));
}
