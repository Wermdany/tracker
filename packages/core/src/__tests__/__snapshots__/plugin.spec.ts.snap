// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PluginDriver inside Api getSortedValidatePlugins 1`] = `"Error running plugin hook \\"build\\" for plugin \\"test\\", expected a function hook or an object with a \\"handler\\" function."`;

exports[`test PluginDriver inside Api runHookSync error 1`] = `"record is not defined"`;
